lib/cretonne/meta/cdsl/ti.py

```diff
@@ -0,0 +1,556 @@
+    """
+    def __init__(self, arg=None):
+        # type: (Optional[Tuple[TypeMap, ConstraintList]]) -> None
+        self.ranks = {}  # type: Dict[TypeVar, int]
+        self.vars = set()  # type: Set[Var]
+
+        if arg is None:
+            self.type_map = {}  # type: TypeMap
+            self.constraints = []  # type: ConstraintList
+        else:
+            self.type_map, self.constraints = arg
+
+        self.idx = 0
+
+    def __getitem__(self, arg):
+        # type: (Union[TypeVar, Var]) -> TypeVar
+        """
+        Lookup the canonical representative for a Var/TypeVar.
+        """
+        if (isinstance(arg, Var)):
+            tv = arg.get_typevar()
+        else:
+            assert (isinstance(arg, TypeVar))
+            tv = arg
+
+        while tv in self.type_map:
+            tv = self.type_map[tv]
+
+        if tv.is_derived:
+            tv = TypeVar.derived(self[tv.base], tv.derived_func)
+        return tv
+
+    def equivalent(self, tv1, tv2):
+        # type: (TypeVar, TypeVar) -> None
+        """
+        Record a that the free tv1 is part of the same equivalence class as
+        tv2.  The canonical representative of the merged class is tv2's
+        cannonical representative.
+        """
+        assert not tv1.is_derived
+        assert self[tv1] == tv1
+
+        # Make sure we don't create cycles
+        if tv2.is_derived:
+            assert self[tv2.base] != tv1
+
+        self.type_map[tv1] = tv2
+
+    def add_constraint(self, tv1, tv2):
+        # type: (TypeVar, TypeVar) -> None
+        """
+        Add a new equivalence constraint between tv1 and tv2
+        """
+        self.constraints.append((tv1, tv2))
+
+    def get_uid(self):
+        # type: () -> str
+        r = str(self.idx)
+        self.idx += 1
+        return r
+
+    def __repr__(self):
+        # type: () -> str
+        return self.dot()
+
+    def rank(self, tv):
+        # type: (TypeVar) -> int
+        """
+        Get the rank of tv in the partial order. TVs directly associated with a
+        Var get their rank from the Var (see register()).
+        Internally generated non-derived TVs implicitly get the lowest rank (0)
+        Internal derived variables get the highest rank.
+        """
+        default_rank = 5 if tv.is_derived else 0
+        return self.ranks.get(tv, default_rank)
+
+    def register(self, v):
+        # type: (Var) -> None
+        """
+        Register a new Var v.  This computes a rank for the associated TypeVar
+        for v, which is used to impose a partial order on type variables.
+        """
+        self.vars.add(v)
+
+        if v.is_input():
+            r = 4
+        elif v.is_intermediate():
+            r = 3
+        elif v.is_output():
+            r = 2
+        else:
+            assert(v.is_temp())
+            r = 1
+
+        self.ranks[v.get_typevar()] = r
+
+    def free_typevars(self):
+        # type: () -> Set[TypeVar]
```

Comment:
Be careful here that we don't introduce a dependency on Python's set ordering. The result of type checking should not depend on it.